pallets/pos-switch/src/lib.rs

```diff
@@ -0,0 +1,83 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::{pallet_prelude::*, SetCode};
+	use sp_core::U256;
+	use sp_std::vec::Vec;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_difficulty::Config {
+		type RuntimeBlockNumber: IsType<<Self as frame_system::Config>::BlockNumber>
+			+ Clone
+			+ Into<sp_core::U256>;
+		/// Because this pallet emits events, it depends on the runtime's definition of an event.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn switch_block_number)]
+	pub type SwitchBlockNumber<T> = StorageValue<_, U256, OptionQuery>;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	// Pallets use events to inform users when important changes are made.
+	// https://substrate.dev/docs/en/knowledgebase/runtime/events
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Switched to PoS. []
+		Switched,
+	}
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		pub switch_block_number: Option<T::BlockNumber>,
+	}
+
+	#[cfg(feature = "std")]
+	impl<T: Config> Default for GenesisConfig<T> {
+		fn default() -> Self {
+			Self { switch_block_number: None }
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {
+			if let Some(switch_block_number) = self.switch_block_number {
+				let switch_block_number = T::RuntimeBlockNumber::from_ref(&switch_block_number);
+				let switch_block_number: sp_core::U256 = switch_block_number.clone().into();
+				SwitchBlockNumber::<T>::put(switch_block_number);
+			}
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Switch to PoS
+		#[pallet::call_index(0)]
+		#[pallet::weight((T::BlockWeights::get().max_block, DispatchClass::Operational))]
+		pub fn switch_to_pos(origin: OriginFor<T>, code: Vec<u8>) -> DispatchResultWithPostInfo {
+			ensure_root(origin)?;
+
+			let block_number = frame_system::Pallet::<T>::block_number();
+			let block_number: sp_core::U256 =
+				T::RuntimeBlockNumber::from_ref(&block_number).clone().into();
+			SwitchBlockNumber::<T>::put(block_number);
+
+			frame_system::Pallet::<T>::can_set_code(&code)?;
```

Comment:
Q: Is the `code` argument the WASM runtime or a block hash where a new runtime with consensus switch was written on-chain ?

Reply:
The WASM runtime. This extrinsic is basically just `System.set_code` plus some extras.

Reply:
A few more questions here: 

0) I am assuming `switch_to_pos()` will trigger migrations then (by extension of calling `set_code`).  Is that correct?

1) Should we update https://github.com/gluwa/creditcoin/blob/dev/scripts/js/src/runtimeUpgrade.ts#L118 ?

2) Do I understand correctly that we need to call `switch_to_pos()` only once and then it's not needed? 

3) What happens after `switch_to_pos()` is already called as part of scheduling the upgrade to mainnet ? Do we just remove the `pos-switch` pallet ? It sounds like we don't need it afterwards.

Reply:
Even more questions:

For 0 and 1) - I think I'll need to modify `runtimeUpgrade.ts` anyway, b/c that's the entry-point for fork-and-migrate tests. Will see how that goes. 

About docker images:

1) Is consensus engine a client side or a runtime component ? I think it is predominantly client side.

2) Assume we upgrade our own nodes with a newer docker image, then call `switch_to_pos()` with a newer runtime. What happens with all of the other nodes which haven't updated to the latest docker image? 

  -- will they still be able to keep in sync with mainnet ? I think yes.
  -- they should definitely not be able to produce blocks b/c the consensus engine has changed. Will these nodes start a fork in the blockchain (a lot of them should be peer connected). Are you aware of any negative consequences that may happen ? 
  -- what will a node operator observe if they aren't aware that consensus has changed  and they haven't upgraded their docker images ?
  -- does the versioning protection from Substrate kick in in this case and how exactly is it going to work? Or will the versioning protection even matter b/c of the consensus change ? 


For 2) and 3) - captured this in CSUB-494.

Reply:
1. Yes, primarily client side
2. Nodes that haven't updated their client will not be able to keep in sync, because syncing the blocks requires having the newer consensus to verify them. This is still, in essence, a hard fork. In theory, they could continue to produce PoW blocks and fork off. To try to avoid that, the `switch_to_pos` extrinsic sets the difficulty to the maximum value which should make producing a PoW block close to impossible.

To a node operator, it will just appear like there are no new blocks until they update their client.

There isn't much version protection in this case, as consensus lives outside the runtime and substrate doesn't actually support switching consensus mid-chain. The ability to switch on the fly is what I've added here (and in our substrate fork) and is the "history preservation" feature.